### PR TITLE
Introduce ComponentContext to Component interface

### PR DESCRIPTION
**Pull request by michael-cico**
# Description

Refactor Component interface calls around new ComponentContext interface
- Collect parameters into a single object to avoid proliferation of parameters
- Context object will eventually own construction of effective (merged) CR
- Update impls and unit tests
- Unrelated, skip empty values files passed into helm upgrade
- Use fake component impls for controller testing

# Checklist 

As the author of this PR, I have:

- [x] Checked that I included or updated copyright and license notices in all files that I altered
- [x] Added or updated unit tests for any new functions I added
- [ ] Added or updated integration tests if appropriate
- [ ] Added or updated acceptance tests if appropriate

Code reviewer, please confirm this PR:

- [ ] Addressed the requirement and meets the acceptance criteria
- [ ] Does not introduce unrelated or spurious changes
- [ ] Does not introduce any unapproved dependency
- [ ] Makes sense and it easy to understand, and/or difficult areas of code are clearly documented so that they can be understood

### CODE REVIEW

File: platform-operator/controllers/verrazzano/component/spi/component.go [hunk 1]
```diff
@@ -4,10 +4,24 @@
...
 )
 
+// ComponentContext Defines the context objects required for Component operations
+type ComponentContext struct {
+	// Client Kubernetes client
+	Client clipkg.Client
+	// DryRun If true, do a dry run of operations
+	DryRun bool
+	// Config Represents the current Verrazzano object state
+	Config *vzapi.Verrazzano
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/component/spi/component.go:19]: change to "verrazzanoCR"

File: platform-operator/controllers/verrazzano/component/spi/component.go [hunk 2]
```diff
@@ -4,10 +4,24 @@
...
+	// EffectiveConfig Represents the actual configuration resulting from the named profiles
+	// used and any configured overrides
+	EffectiveConfig *vzapi.Verrazzano
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/component/spi/component.go:22]: mergedCR ?

File: platform-operator/controllers/verrazzano/component/spi/component.go [hunk 3]
```diff
@@ -4,10 +4,24 @@
...
 
 import (
+	vzapi "github.com/verrazzano/verrazzano/platform-operator/apis/verrazzano/v1alpha1"
 	"go.uber.org/zap"
 	clipkg "sigs.k8s.io/controller-runtime/pkg/client"
 )
 
+// ComponentContext Defines the context objects required for Component operations
+type ComponentContext struct {
+	// Client Kubernetes client
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/component/spi/component.go:14]: Add log in the context

File: platform-operator/controllers/verrazzano/component/spi/component.go [hunk 4]
```diff
@@ -4,10 +4,24 @@
...
+	DryRun bool
+	// Config Represents the current Verrazzano object state
+	Config *vzapi.Verrazzano
+	// EffectiveConfig Represents the actual configuration resulting from the named profiles
+	// used and any configured overrides
+	EffectiveConfig *vzapi.Verrazzano
+}
+
 // Component interface defines the methods implemented by components
 type Component interface {
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/component/spi/component.go:26]: Break up into install, upgrade, update interfaces.

File: platform-operator/controllers/verrazzano/component/spi/component.go (hunk 3 above)
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/spi/component.go:14]: ✅

File: platform-operator/controllers/verrazzano/component/spi/component.go (hunk 1 above)
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/spi/component.go:19]: Changed to "CR" and "EffectiveCR"

**Review by markxnelson — APPROVED** [review]

File: platform-operator/controllers/verrazzano/component/spi/component.go [hunk 5]
```diff
@@ -4,47 +4,68 @@
...
 	// Name returns the name of the Verrazzano component
 	Name() string
+	// GetDependencies returns the dependencies of this component
+	GetDependencies() []string
+	// IsReady Indicates whether or not a component is available and ready
+	IsReady(context ComponentContext) bool
+}
 
+// InstallComponent interface defines installs operations for components that support it
+type InstallComponent interface {
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/component/spi/component.go]: The convention is use use -er,   such as ComponentInstaller,   ComponentUpgrader, etc.  Up to you if you want to change it.  https://golang.org/doc/effective_go#interface-names

Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/spi/component.go]: Fixed and pushed

**Review by pfmackin — APPROVED** [review]

**Pull request merged**